### PR TITLE
New PWM implementation

### PR DESCRIPTION
**Pull request by csagonero**
Support PC6/PC7/PC8/PC9 and PB4/PB5 TIMER3

Signed-off-by: Cyril SAGONÉRO <csagonero@free.fr>

### CODE REVIEW

File: stm32f0_hal/examples/pwm.rs [hunk 1]
```diff
@@ -3,11 +3,22 @@
...
-    pwm::init(10000);
-    pwm::set_duty(750);
-    pwm::enable();
-
+    rcc::init();
+    let channel1 = pwm::Pwm::init(pwm::Pin::PC6);
+    let channel2 = pwm::Pwm::init(pwm::Pin::PC7);
+    let channel3 = pwm::Pwm::init(pwm::Pin::PC8);
+    let channel4 = pwm::Pwm::init(pwm::Pin::PC9);
+    channel1.set_frequency(49);
```
Comment (pierre-rouanet) [inline stm32f0_hal/examples/pwm.rs:14]: It's weird that you are only setting the frequency for the channel 1.

File: stm32f0_hal/src/pwm.rs [hunk 2]
```diff
@@ -1,73 +1,190 @@
...
 use cortex_m;
 
-// Init Timer Function without remap pin
-pub fn init(period: u32) {
-    cortex_m::interrupt::free(|cs| {
-        let rcc = RCC.borrow(cs);
-        let tim3 = TIM3.borrow(cs);
-        let gpiob = GPIOB.borrow(cs);
+static mut FREQ: u32 = 0;
+static mut DUT: u8 = 0;
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/pwm.rs]: DUTY?

File: stm32f0_hal/src/pwm.rs [hunk 3]
```diff
@@ -1,73 +1,190 @@
...
-            w.cms()
-                .center_align_mode1()
-                .dir()
-                .up()
-                .opm()
-                .continuous()
-                .cen()
-                .enabled()
-        });
+            // GPIOB Clock Activated
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/pwm.rs]: and GPIOC

File: stm32f0_hal/src/pwm.rs [hunk 4]
```diff
@@ -1,73 +1,190 @@
...
+            let tim3 = TIM3.borrow(cs);
 
-pub fn enable() {
-    cortex_m::interrupt::free(|cs| {
-        let tim3 = TIM3.borrow(cs);
-        tim3.ccer.modify(|_, w| w.cc1e().set_bit());
-    });
-}
+            // Set Prescaler Register - 16 bits
+            tim3.psc.write(|w| w.psc().bits(23)); // 1 MHz // 23
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/pwm.rs]: Can you explain this 23?

File: stm32f0_hal/src/pwm.rs [hunk 5]
```diff
@@ -1,73 +1,190 @@
...
+                Pin::PC8 => tim3.ccer.modify(|_, w| w.cc3e().set_bit()),
+                Pin::PC9 => tim3.ccer.modify(|_, w| w.cc4e().set_bit()),
+            }
+        });
+    }
+
+    pub fn set_duty(&self, duty: u8) {
+        cortex_m::interrupt::free(|cs| {
+            let tim3 = TIM3.borrow(cs);
+            unsafe {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/pwm.rs]: Could we make this unsafe only about setting DUT?

File: stm32f0_hal/src/pwm.rs [hunk 6]
```diff
@@ -1,73 +1,190 @@
...
+        Pwm { pin }
+    }
 
-        // Set Auto-Reload register - 32 bits
-        let arr = 10000;
-        tim3.arr.write(|w| w.arr_h().bits((arr >> 16) as u16));
-        tim3.arr.write(|w| w.arr_l().bits(arr as u16));
-    });
-}
+    pub fn set_frequency(&self, frequency: u32) {
```
Comment (pierre-rouanet) [inline stm32f0_hal/src/pwm.rs]: If I understand correctly, this set_frequency actually impact all PWM instances? 
Maybe it's better to make it a class method and for the moment add a comment specifying that you should reset all duty of your instances. What do you think?

File: stm32f0_hal/examples/pwm.rs (hunk 1 above)
Comment (csagonero) [reply stm32f0_hal/examples/pwm.rs:14]: In fact, you have only one frequency for all the channel of the same TIMER.

File: stm32f0_hal/src/pwm.rs (hunk 3 above)
Comment (csagonero) [reply stm32f0_hal/src/pwm.rs]: ok

**Pull request merged**